azure/services/privatedns/spec.go

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package privatedns
+
+import (
+	"context"
+
+	"github.com/Azure/azure-sdk-for-go/services/privatedns/mgmt/2018-09-01/privatedns"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/pkg/errors"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+	"sigs.k8s.io/cluster-api-provider-azure/azure/converters"
+	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
+)
+
+// ZoneSpec defines the specification for private dns zone.
+type ZoneSpec struct {
+	Name           string
+	ResourceGroup  string
+	ClusterName    string
+	AdditionalTags infrav1.Tags
+}
+
+// LinkSpec defines the specification for a virtual network link in a private DNS zone.
+type LinkSpec struct {
+	Name              string
+	ZoneName          string
+	SubscriptionID    string
+	VNetResourceGroup string
+	VNetName          string
+	ResourceGroup     string
+	ClusterName       string
+	AdditionalTags    infrav1.Tags
+}
+
+type RecordSpec struct {
+	Record        infrav1.AddressRecord
+	ZoneName      string
+	ResourceGroup string
+}
+
+// ResourceName returns the name of the private dns zone.
+func (z ZoneSpec) ResourceName() string {
+	return z.Name
+}
+
+// OwnerResourceName is a no-op for private dns zone.
+func (z ZoneSpec) OwnerResourceName() string {
+	return ""
+}
+
+// ResourceGroupName returns the name of the resource group of the private dns zone.
+func (z ZoneSpec) ResourceGroupName() string {
+	return z.ResourceGroup
+}
+
+// Parameters returns the parameters for the private dns zone.
+func (z ZoneSpec) Parameters(existing interface{}) (params interface{}, err error) {
+	_, log, done := tele.StartSpanWithLogger(context.TODO(), "privatedns.ZoneSpec.Parameters")
+	defer done()
+
+	if existing != nil {
+		zone, ok := existing.(privatedns.PrivateZone)
+		if !ok {
+			return nil, errors.Errorf("%T is not a privatedns.PrivateZone", existing)
+		}
+		if !isManaged(zone.Tags, z.ClusterName) {
+			log.V(1).Info("Skipping reconciliation of unmanaged private DNS zone", "private DNS", zone.Name)
+			log.V(1).Info("Tag the DNS manually from azure to manage it with capz."+
+				"Please see https://capz.sigs.k8s.io/topics/custom-dns.html#manage-dns-via-capz-tool", "private DNS", zone.Name)
+			return nil, nil
+		}
```

Comment:
Should we return `nil, nil` at the end of this block anyway if the resource exists? If that's the case, then the check for `isManaged` would be redundant.

Reply:
I thought about this, but didn't want to change the current implementation. wdyt about doing it in a followup pr?

Reply:
I think it should be fine in either case since if we don't update, we'd just be returning the existing resource anyway. In general, I've set up Parameters() to return nil, nil if we find an existing resource and also added any checks to the fields, i.e. for a subnet, if the VNet exists when it's not managed.

Reply:
yeah, we [discussed](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1791#discussion_r749571484) about this earlier and created a followup for it. I'm fine with doing it here; I left it for now because we want to focus on other things. Happy to do it either way.

Reply:
Okay sounds good.